compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs

```diff
@@ -559,13 +563,118 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        // Suggest:
+        //  - Changing the return type to be `impl <all bounds>`
+
+        debug!("try_suggest_return_impl_trait, expected = {:?}, found = {:?}", expected, found);
+
+        let ty::Param(expected_ty_as_param) = expected.kind() else { return };
+
+        let fn_node = self.tcx.hir().find(fn_id);
+
+        let Some(hir::Node::Item(hir::Item {
+            kind:
+                hir::ItemKind::Fn(
+                    hir::FnSig { decl: hir::FnDecl { inputs: fn_parameters, output: fn_return, .. }, .. },
+                    hir::Generics { params, where_clause, .. },
+                    _body_id,
+                ),
+            ..
+        })) = fn_node else { return };
+
+        let Some(expected_generic_param) = params.get(expected_ty_as_param.index as usize) else { return };
+
+        // get all where BoundPredicates here, because they are used in to cases below
+        let where_predicates = where_clause.predicates.iter().filter_map(|p| match p {
+            WherePredicate::BoundPredicate(hir::WhereBoundPredicate {
+                bounds, bounded_ty, ..
+            }) => {
+                let ty = <dyn AstConv<'_>>::ast_ty_to_ty(self, bounded_ty);
+                Some((ty, bounds))
+            }
+            _ => None,
+        });
+
+        // check whether there is any predicate that contains our `T`, like `Option<T>: Send`
+        let any_predicate_contains_expected_ty =
+            where_predicates.clone().any(|(ty, _)| match ty.kind() {
+                ty::Param(param_ty) if param_ty == expected_ty_as_param => false,
+                _ => ty.contains(expected, self.tcx),
+            });
+
+        if any_predicate_contains_expected_ty {
+            return;
+        }
+
+        // now get all predicates in the same types as the where bounds, so we can chain them
+        let predicates_from_where = where_predicates
+            .filter_map(|(ty, bounds)| match ty.kind() {
+                ty::Param(param_ty) if param_ty == expected_ty_as_param => Some(bounds),
+                _ => None,
+            })
+            .map(|bounds| bounds.iter())
+            .flatten();
+
+        // extract all bounds from the source code using their spans
+        let all_matching_bounds_strs = expected_generic_param
+            .bounds
+            .iter()
+            .chain(predicates_from_where)
+            .filter_map(|bound| match bound {
+                GenericBound::Trait(_, _) => {
+                    self.tcx.sess.source_map().span_to_snippet(bound.span()).ok()
+                }
+                _ => None,
+            })
+            .collect::<Vec<String>>();
+
+        if all_matching_bounds_strs.len() != 0 {
+            let all_bounds_str = all_matching_bounds_strs.join(" + ");
+
+            let ty_param_used_in_fn_params = fn_parameters.iter().any(|param| {
+                let ty = <dyn AstConv<'_>>::ast_ty_to_ty(self, param);
```

Comment:
Did you get a chance to look at the `ast_ty_to_ty` calls? I'd at least add a FIXME, even if you don't change this now.